pkg/lib/operation/project/local/template/upgrade/operation.go

```diff
@@ -25,6 +36,163 @@ type dependencies interface {
+		if tmplObjectState.Kind().IsBranch() {
+			continue
+		}
+
+		var opMark string
+		var objectState model.ObjectState
+		if v, found := projectState.Get(tmplObjectState.Key()); found {
+			opMark = diff.ChangeMark
+			objectState = v
+			objectState.SetLocalState(tmplObjectState.LocalState())
+
+			// Clear path
+			objectState.Manifest().SetParentPath("")
+			objectState.Manifest().SetRelativePath("")
+			renameOp.Add(objectState)
+		} else {
+			opMark = diff.AddMark
+			objectState = tmplObjectState
+
+			// Clear path
+			objectState.Manifest().SetParentPath("")
+			objectState.Manifest().SetRelativePath("")
+
+			// Copy state from template to project
+			if err := projectState.Set(objectState); err != nil {
+				errors.Append(err)
+				continue
+			}
+
+			// Generate path
+			renameOp.Add(objectState)
+		}
+
+		objects = append(objects, upgradedObject{ObjectState: objectState, opMark: opMark})
+
+		// Save to filesystem
+		saveOp.SaveObject(objectState, objectState.LocalState(), model.NewChangedFields())
+	}
+
+	// Delete
+	var toDelete []model.Key
+	configs := search.ConfigsForTemplateInstance(projectState.LocalObjects().ConfigsWithRowsFrom(o.Branch), o.Instance.InstanceId)
+	for _, config := range configs {
+		if _, found := templateState.Get(config.Key()); !found {
+			toDelete = append(toDelete, config.Key())
+		}
+		for _, row := range config.Rows {
+			if _, found := templateState.Get(row.Key()); !found {
+				toDelete = append(toDelete, row.Key())
+			}
+		}
+	}
+	for _, key := range toDelete {
+		objectState := projectState.MustGet(key)
+		saveOp.DeleteObject(objectState, objectState.Manifest())
+		objects = append(objects, upgradedObject{ObjectState: objectState, opMark: diff.DeleteMark})
+	}
+
+	// Execute rename and save
+	if errors.Len() > 0 {
+		return errors
+	}
+	if err := renameOp.Invoke(); err != nil {
+		return err
+	}
+	if err := saveOp.Invoke(); err != nil {
+		return err
+	}
+
+	// Encrypt values
+	if err := encrypt.Run(projectState, encrypt.Options{DryRun: false, LogEmpty: false}, d); err != nil {
+		return err
+	}
+
+	// Save manifest
+	if _, err := saveProjectManifest.Run(projectState.ProjectManifest(), projectState.Fs(), d); err != nil {
+		return err
+	}
+
+	// Log new objects
+	objects.Log(logger, tmpl)
+
+	// Normalize paths
+	if _, err := rename.Run(projectState, rename.Options{DryRun: false, LogEmpty: false}, d); err != nil {
+		return err
+	}
+
+	// Validate schemas and encryption
+	if err := validate.Run(projectState, validate.Options{ValidateSecrets: true, ValidateJsonSchema: true}, d); err != nil {
+		logger.Warn(`Warning, ` + err.Error())
+		logger.Warn()
+		logger.Warnf(`Please correct the problems listed above.`)
+		logger.Warnf(`Push operation is only possible when project is valid.`)
+	}
+
+	logger.Info(fmt.Sprintf(`Template instance "%s" has been upgraded to "%s".`, o.Instance.InstanceId, tmpl.FullName()))
+	return nil
+}
+
+type upgradedObject struct {
+	model.ObjectState
+	opMark string
```

Comment:
`add/change/delete` ... neviem spravit diff (co opatovne aplikovanie sablony upravilo), tak som to vyriesil docasne takto.

Ak chce uzivatel vidiet presne zmeny, tak moze zavolat `kbc diff`, ... rovnako ich bude vidiet v `kbc push`.
Vysledkom tohto kompromisu je, ze vsetky pregenerovane objekty, maju `mark` - `* - changed`, ... bez ohladu na to, ci sa pregenerovanim zmenili alebo nie.

Reply:
Vid E2E test.